File: desktop/src/agent-core/planner.ts
import { ChatOpenAI } from "@langchain/openai";
import { ChatPromptTemplate } from "@langchain/core/prompts";
import { StructuredOutputParser } from "@langchain/core/output_parsers";
import { z } from "zod";
import * as dotenv from "dotenv";
import { ToolRegistry } from "./mcp/registry";
import { ToolCall } from "./mcp/schemas";

dotenv.config();

const StepSchema = z.object({
    id: z.number(),
    action: z.enum(["create", "modify", "delete", "run_command", "review", "tool_call"]),
    path: z.string().optional().default(""),
    description: z.string(),
    content: z.string().optional(),
    tool_call: z.custom<ToolCall>().optional(),
    status: z.enum(["pending", "running", "completed", "failed"]).default("pending"),
});

const PlanSchema = z.object({
    goal: z.string(),
    steps: z.array(StepSchema),
});

export type Step = z.infer<typeof StepSchema>;
export type Plan = z.infer<typeof PlanSchema>;

export class Planner {
    private registry: ToolRegistry;

    constructor() {
        this.registry = ToolRegistry.getInstance();
    }

    async createPlan(goal: string): Promise<Plan> {
        const apiKey = process.env.OPENAI_API_KEY;

        if (apiKey && apiKey.startsWith("sk-")) {
            try {
                const llm = new ChatOpenAI({
                    openAIApiKey: apiKey,
                    modelName: "gpt-4o",
                    temperature: 0,
                });

                const parser = StructuredOutputParser.fromZodSchema(PlanSchema);

                const tools = this.registry.listTools();
                const toolsDesc = tools.map(t => `- ${t.name}: ${t.description}`).join("\n");

                const systemPrompt = `You are an expert AI software architect.
Create a detailed, step-by-step implementation plan.

Available Tools:
${toolsDesc}

For each step, specify:
- action: 'tool_call' (PREFERRED) or legacy actions ('create', 'modify', 'run_command', 'review')
- tool_call: If action is 'tool_call', provide the tool name and arguments.
- description: brief explanation of what to do
- path: if applicable (e.g. for file operations)

Ensure dependencies are handled first.`;

                const prompt = ChatPromptTemplate.fromMessages([
                    ["system", systemPrompt],
                    ["user", "{goal}\n\n{format_instructions}"],
                ]);

                const chain = prompt.pipe(llm).pipe(parser);

                const result = await chain.invoke({
                    goal: goal,
                    format_instructions: parser.getFormatInstructions(),
                });

                return result as Plan;
            } catch (error) {
                console.warn(`[Warning] LLM planning failed: ${error}. Falling back to mock.`);
            }
        } else {
            console.info("[Info] No OpenAI API key found. Using mock planner.");
        }

        // Mock Plan Fallback
        return {
            goal: goal,
            steps: [
                {
                    id: 1,
                    action: "create",
                    path: "desktop/src/test-mock.ts",
                    content: "console.log('Hello from Node Mock Planner')",
                    description: "Create a test file",
                    status: "pending"
                },
                {
                    id: 2,
                    action: "run_command",
                    path: "",
                    description: "Run a safe command",
                    status: "pending"
                },
            ],
        };
    }
}

File: core/planner.py
from typing import List, Dict, Any, Optional
from pydantic import BaseModel
import os
from dotenv import load_dotenv
from langchain_core.prompts import ChatPromptTemplate
from langchain_openai import ChatOpenAI
from langchain_core.output_parsers import PydanticOutputParser
from .tools.mcp.registry import ToolRegistry
from .tools.builtins.filesystem import create_filesystem_client
from .tools.integrations.github import GitHubMCPClient
from .tools.mcp.schemas import ToolCall

load_dotenv()

class Step(BaseModel):
    id: int
    action: str = "review" # create, modify, delete, run_command, review, tool_call
    path: str = ""
    description: str
    content: Optional[str] = None # New content for create/modify
    tool_call: Optional[ToolCall] = None # Structured tool call
    status: str = "pending"

class Plan(BaseModel):
    goal: str
    steps: List[Step]

class Planner:
    def __init__(self):
        self.registry = ToolRegistry()
        self._register_default_tools()

    def _register_default_tools(self):
        # 1. FileSystem (sandboxed)
        fs_client = create_filesystem_client()
        self.registry.register_server("filesystem", fs_client)
        for tool in fs_client.list_tools():
            self.registry.register_tool("filesystem", tool)

        # 2. GitHub
        gh_client = GitHubMCPClient()
        self.registry.register_server("github", gh_client)
        for tool in gh_client.list_tools():
            self.registry.register_tool("github", tool)

    def create_plan(self, goal: str) -> Plan:
        """
        Generates a plan for the given goal using an LLM if available, otherwise returns a mock.
        """
        api_key = os.getenv("OPENAI_API_KEY")
        
        if api_key and api_key.startswith("sk-"):
            try:
                llm = ChatOpenAI(api_key=api_key, model="gpt-4o", temperature=0)
                parser = PydanticOutputParser(pydantic_object=Plan)
                
                tools_desc = "\n".join([f"- {t.name}: {t.description}" for t in self.registry.list_tools()])
                
                system_prompt = f"""You are an expert AI software architect.
                Create a detailed, step-by-step implementation plan.
                
                Available Tools:
                {tools_desc}
                
                For each step, specify:
                - action: 'tool_call' (PREFERRED) or legacy actions ('create', 'modify', 'run_command')
                - tool_call: If action is 'tool_call', provide the tool name and arguments.
                - description: brief explanation of what to do
                
                Ensure dependencies are handled first.
                """
                
                prompt = ChatPromptTemplate.from_messages([
                    ("system", system_prompt),
                    ("user", "{goal}\n\n{format_instructions}")
                ])
                
                chain = prompt | llm | parser
                return chain.invoke({"goal": goal, "format_instructions": parser.get_format_instructions()})
            except Exception as e:
                print(f"[Warning] LLM planning failed: {e}. Falling back to mock.")
        else:
            print("[Info] No OpenAI API key found. Using mock planner.")

        # Mock Plan Fallback (Safe for testing)
        mock_steps = [
            Step(id=1, action="create", path="core/test_mock.py", content="print('Hello form Mock Planner')", description="Create a test file"),
            Step(id=2, action="run_command", path="echo 'Mock execution running'", description="Run a safe command"),
        ]
        
        return Plan(goal=goal, steps=mock_steps)
